### PR TITLE
feat(ast) Expand alias references at the end of parsing to hide them to query processing

### PR DESCRIPTION
**Pull request by fpacifici**
Stacked on top of #1023 

The query processing in Snuba (both for old representation and new representation) expects to work transparently to aliases references. The query that is processed is fully inlined and then it is packaged back at the end before formatting when alias references are reintroduced. This is from the times of column_expr/old prewhere processor that would transform columns into expressions like `nullIf(group_id, 0) as group_id` in several places around the query. So an alias is defined multiple times with the same expression and optimized at the end where it is added to the clickhouse expression only once (the other occurrences  are references to that aliases).

... Or maybe not ...
The issue is that the query can have references to alias defined in the query itself. These references are never expanded and linger as references during query processing, which at that point likely does not behave properly (that's probably the reason why they are not extensively used apart for simple cases queries where they happen to work) since methods like get_all_columns_in_query/condition would not behave properly and even column_expr may erroneously transform an alias reference believing it is a column.

This can be worse with the AST since it unlocks more complex processing.

Assuming we keep such design where the query is fully expanded and inlined during query processing and we enforce that no query processor introduces alias shadowing (easy to do), to make this model consistent we need to expand alias references at the end of parsing (so they disappear) and are reintroduced at the end before formatting.
That's what this PR does:
- it finds all aliases declared in the query
- it replaces alias references with the expanded expressions and as a bonus enforces no cycles are created (which before would have been rejected by clickhouse).

The way query processing works does not change. A query processor decides whether they want to just apply a pure transformation function to the whole query (which is by far the most common case and where they structurally cannot introduce alias shadowing) or, for more advanced  processing they have to mindful of aliases.

The open point that I will address by adding monitoring in another PR is whether there are use cases in production that exploit the flaw of the current model, cases where the query works because query processing fails at dealing with alias references.

A potential followup is to revisit whether the model where query processing is performed in a fully expanded query vs we keep references nodes around during query processing (there are drawbacks and advantages for both). I am scheduling a design meeting to discuss that specific problem. The scope here is to fix the current model and make it work as it is supposed to.
The fact that the AST is not yet in production gives us the opportunity to make the model consistent with itself at least and to rule out production cases that exploit the flaw to work.

### CODE REVIEW

File: snuba/query/parser/__init__.py [hunk 1]
```diff
@@ -200,3 +216,181 @@ def apply_aliases(exp: Expression) -> Expression:
...
+    """
+    Recursively expand all the references to aliases in the query. This
+    makes life easy to query processors and translators that only have to
+    take care of not introducing shadowing (easy to enforce at runtime),
+    otherwise aliasing is transparent to them.
+    """
+    # Pre-inline all nested aliases. This reduces the number of iterations
+    # we need to do on the query.
+    # Example:
+    # {"a": f(x), "x": g(k)} -> {"a": f(g(k)), "x": g(k)}
```
Comment (fpacifici) [inline snuba/query/parser/__init__.py:231]: This is just an optimization to do fewer scans over the query that does not compromise readability to me. When there is a nested expression that defines an alias and contains references to other aliases, if we pre-expand that expression enttirely, when we expand the references in the query, the expression is already fully expanded and we do not need to expand it again each time.

File: snuba/query/parser/__init__.py [hunk 2]
```diff
@@ -224,3 +240,186 @@ def apply_aliases(exp: Expression) -> Expression:
...
+                lambda clause: replace(
+                    clause, expression=clause.expression.accept(visitor)
+                ),
+                query.get_orderby_from_ast(),
+            )
+        )
+    )
+
+
+class AliasExpanderVisitor(ExpressionVisitor[Expression]):
```
Comment (fpacifici) [inline snuba/query/parser/__init__.py:294]: See tests/query/parser/test_resolver_visitor.py for a coverage of the corner cases

File: snuba/query/parser/__init__.py [hunk 3]
```diff
@@ -224,3 +240,186 @@ def apply_aliases(exp: Expression) -> Expression:
             return replace(exp, alias=exp.column_name)
 
     query.transform_expressions(apply_aliases)
+
+
+def _expand_aliases(query: Query) -> None:
```
Comment (fpacifici) [inline snuba/query/parser/__init__.py:245]: See https://github.com/getsentry/snuba/pull/1024/files#diff-dd1e177b6b14ec2cdec36b243782b920R200 to see how this works in practice.

File: tests/test_split.py [hunk 4]
```diff
@@ -260,27 +260,6 @@ def do_query(
         },
         False,
     ),  # Valid query but not enough columns to split.
-    (
```
Comment (fpacifici) [inline tests/test_split.py:263]: This condition now cannot happen since we are inlining aliases. The result is that the column split query will always be valid even if we remove expressions that were defining aliases.

**Review by tkaemming — APPROVED** [review]

File: tests/test_split.py (hunk 4 above)
Comment (fpacifici) [reply tests/test_split.py:263]: Original case:
the query was something like:
```
SELECT ... stuff .... transform (an ugly expression) AS projects
ORDER BY projects
```

The column splitter would build this query out of that 
```
SELECT project, event_id, timestamp, 
ORDER BY projects
```
`Projects` now is not declared and the query fails. (old and new system)

So I introduced a check in the column splitter that ensured the transformed query was still semantically valid (at least in terms of columns requested and aliases).
Now that we are expanding the aliases I was not able anymore to make the column splitter generate a broken query, since the aliased expression is preserved.

File: tests/query/parser/test_resolver_visitor.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    ),
+    pytest.param(
+        Literal(None, "a"),
+        {},
+        False,
+        Literal(None, "a"),
+        id="Simple Literal - do nothing",
+    ),
+    pytest.param(
+        Column(None, None, "ref"),
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply tests/query/parser/test_resolver_visitor.py]: I don't have a very strong opinion here, and this is ultimately a stylistic call so I'll leave it up to your judgement.

I think that my specific issues are mostly around that I still haven't internalized that the alias is read first so I find myself having to read these multiple times to interpret the structure correctly and in those cases the explicit `None` make it seems to imply the value has more importance than it does in practice.

[the remark this answers is not in the document]
Comment (tkaemming) [reply tests/query/parser/test_resolver_visitor.py]: I see what you mean now (as of 63f7c9f) and I think you were right. 😬

[the remark this answers is not in the document]
Comment (fpacifici) [reply tests/query/parser/test_resolver_visitor.py]: I see the point of the aliases being mostly None (which is kind of the goal here since I am testing alias references). I turned it into key value arguments. Hopefully we will be able to make it less verbose after getting rid of aliases in the expressions entirely.

File: snuba/query/parser/__init__.py [hunk 6]
```diff
@@ -224,3 +240,186 @@ def apply_aliases(exp: Expression) -> Expression:
...
+    query.set_ast_orderby(
+        list(
+            map(
+                lambda clause: replace(
+                    clause, expression=clause.expression.accept(visitor)
+                ),
+                query.get_orderby_from_ast(),
+            )
+        )
+    )
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/parser/__init__.py]: moved to the query class. I think it works fine.

File: snuba/query/parser/__init__.py [hunk 7]
```diff
@@ -224,3 +241,167 @@ def apply_aliases(exp: Expression) -> Expression:
...
+            # `f(a) as b, f(b) as a` -> not allowed
+            if self.__visited_stack[-1] != name:
+                # We need to reject conditions like `f(g(a) as b) as a`
+                # but accept `f(g(a)) as a`.
+                # If we are here it means we already found, in this nested
+                # expression, an alias declared that has the same name of
+                # the column we are visiting (shadowing). So the previously
+                # visited alias must be on top of the stack, to be valid.
+                # In `f(g(a) as b) as a`, instead, b would be on top of the
+                # stack instead of a.
```
Comment (tkaemming) [inline snuba/query/parser/__init__.py:322]: This is much clearer. 👍

File: snuba/query/logical.py [hunk 8]
```diff
@@ -169,6 +169,38 @@ def transform_expression_list(
             )
         )
 
+    TVisited = TypeVar("TVisited")
```
Comment (tkaemming) [inline snuba/query/logical.py:172]: I don't think this is used?

File: snuba/query/logical.py [hunk 9]
```diff
@@ -169,6 +169,38 @@ def transform_expression_list(
...
+        if self.__having is not None:
+            self.__having = self.__having.accept(visitor)
+        self.__order_by = list(
+            map(
+                lambda clause: replace(
+                    clause, expression=clause.expression.accept(visitor)
+                ),
+                self.__order_by,
+            )
+        )
```
Comment (tkaemming) [inline snuba/query/logical.py:202]: Yeah, I think this looks really nice.

File: tests/test_split.py (hunk 4 above)
Comment (tkaemming) [reply tests/test_split.py:263]: This makes sense, thanks for clarifying.

File: snuba/query/logical.py (hunk 8 above)
Comment (fpacifici) [reply snuba/query/logical.py:172]: errrr, it was used till 30 minutes ago. Then I forced the visitor to be parametrized on expressions. Will remove in a minute.
